src/main/cpp/src/map_zip_with_utils.cu

```diff
@@ -0,0 +1,359 @@
+  auto const comparator =
+    cudf::experimental::row::equality::two_table_comparator(values_tview, keys_tview, stream);
+  auto const d_comp    = comparator.equal_to<false>(cudf::nullate::DYNAMIC{has_nulls});
+  using lhs_index_type = cudf::experimental::row::lhs_index_type;
+  using rhs_index_type = cudf::experimental::row::rhs_index_type;
+
+  // Perform the actual key-value comparisons
+  // For each comparison: if key == value, return the value index; otherwise return Out of Bounds
+  // index
+  auto compare_results = make_numeric_column(
+    data_type(type_to_id<size_type>()), total_compares, cudf::mask_state::UNALLOCATED, stream);
+  thrust::transform(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator(0),
+    thrust::make_counting_iterator(total_compares),
+    compare_results->mutable_view().template begin<size_type>(),
+    cuda::proclaim_return_type<size_type>(
+      [total_compares,
+       values_idx,
+       d_key_index,
+       d_val_offsets,
+       d_comp,
+       d_value_sizes_val_offsets = search_values.offsets_begin()] __device__(auto const idx) {
+        return d_comp(static_cast<lhs_index_type>(values_idx[idx]),
+                      static_cast<rhs_index_type>(d_key_index[idx]))
+                 ? values_idx[idx] - d_value_sizes_val_offsets[d_val_offsets[idx]]
+                 : -total_compares - 1;
+      }));
+  // Use segmented reduction to find the minimum value (first match) for each list
+  // This gives us the first matching value index for each row, or Out of Bounds index if no match
+  // found
+  auto offsets_span = cudf::device_span<cudf::size_type const>(values_sizes_offsets->view());
+  auto results =
+    cudf::segmented_reduce(compare_results->view(),
+                           offsets_span,
+                           *cudf::make_max_aggregation<cudf::segmented_reduce_aggregation>(),
+                           cudf::data_type{cudf::type_to_id<size_type>()},
+                           cudf::null_policy::EXCLUDE,
+                           stream,
+                           mr);
+  return results;
+}
+
+std::unique_ptr<cudf::column> map_zip(
+  cudf::lists_column_view const& col1,  // First map column containing key-value pairs
+  cudf::lists_column_view const& col2,  // Second map column containing key-value pairs
+  rmm::cuda_stream_view stream,         // CUDA stream for asynchronous execution
+  rmm::device_async_resource_ref mr)    // Memory resource for allocations
+{
+  CUDF_EXPECTS(col1.child().type().id() == cudf::type_id::STRUCT,
+               "col1 must have exactly 1 child (STRUCT) column.");
+  CUDF_EXPECTS(col1.child().num_children() == 2,
+               "col1 key-value struct must have exactly 2 children.");
+  CUDF_EXPECTS(col2.child().type().id() == cudf::type_id::STRUCT,
+               "col2 must have exactly 1 child (STRUCT) column.");
+  CUDF_EXPECTS(col2.child().num_children() == 2,
+               "col2 key-value struct must have exactly 2 children.");
+
```

Comment:
We're making an assumption that neither maps columns have null keys in them.  Spark wouldn't allow it.

It might be good to add an assert that neither keys column `.has_nulls()`.